### PR TITLE
feat: pba payment screen (CMC-1062)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CMC-1062

### Change description ###

PRD client
Dynamic list for PBA accounts
Store fee data in CCD

**Does this PR introduce a breaking change?** (check one with "x")

Removal of old fixed list PBA field will break all exisiting cases in all envs

```
[x] Yes
[] No
```

### CODE REVIEW

File: ccd-definition/CaseField.json [hunk 1]
```diff
@@ -118,6 +118,57 @@
...
+    "FieldType": "Label",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "UNSPECIFIED_CLAIMS",
+    "ID": "applicantSolicitor1PbaAccountsIsEmpty",
+    "Label": "Hide pba and show noPbaAccountLabel",
+    "FieldType": "YesOrNo",
+    "SecurityClassification": "Public",
```
Comment (pd-dudkiewicz) [inline ccd-definition/CaseField.json:159]: There was this conversation a while ago about having `_Category` on case fields. I don't think we agreed on approach for all new fields, but this one could be marked as a `WORKAROUND`.

File: e2e/pages/createClaim/paymentReference.page.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+module.exports = {
+
+  fields: {
+    paymentReference: '#paymentReference',
+  },
+
+  async updatePaymentReference() {
+    I.waitForElement(this.fields.paymentReference);
+    I.clearField(this.fields.paymentReference);
```
Comment (pd-dudkiewicz) [inline e2e/pages/createClaim/paymentReference.page.js:11]: why is this required?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java [hunk 3]
```diff
@@ -74,6 +87,34 @@ private CallbackResponse validateDateOfBirth(CallbackParams callbackParams) {
...
+        Optional<SolicitorReferences> references = ofNullable(caseData.getSolicitorReferences());
+        String paymentReference = ofNullable(caseData.getPaymentReference())
+            .orElse(references.map(SolicitorReferences::getApplicantSolicitor1Reference).orElse(""));
+
+        String authToken = callbackParams.getParams().get(BEARER_TOKEN).toString();
+        List<String> pbaNumbers = getPbaAccounts(authToken);
+
+        CaseData.CaseDataBuilder caseDataBuilder = caseData.toBuilder()
+            .claimFee(feesService.getFeeDataByClaimValue(caseData.getClaimValue()))
+            .applicantSolicitor1PbaAccounts(DynamicList.fromList(pbaNumbers))
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java:110]: We should probably check if there's some value selected, otherwise it will be removed when going to previous page and visiting this one again. Might be useful thing to have in `DynamicList` class.

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java [hunk 4]
```diff
@@ -74,6 +87,34 @@ private CallbackResponse validateDateOfBirth(CallbackParams callbackParams) {
...
+            .data(caseDetailsConverter.toMap(caseDataBuilder.build()))
+            .build();
+    }
+
+    private List<String> getPbaAccounts(String authToken) {
+        Optional<Organisation> organisation = organisationService.findOrganisation(authToken);
+
+        List<String> pbaNumbers = new ArrayList<>();
+        organisation.ifPresent(org -> pbaNumbers.addAll(org.getPaymentAccount()));
+        return pbaNumbers;
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java]: How about:
```
return organisationService.findOrganisation(authToken)
    .map(Organisation::getPaymentAccount)
    .orElse(List.of());
```

File: src/main/java/uk/gov/hmcts/reform/unspec/service/FeesService.java [hunk 5]
```diff
@@ -40,15 +34,13 @@ private FeeLookupResponseDto lookupFee(ClaimValue claimValue) {
...
 
-    private BigInteger getFeeAmountInPence(FeeLookupResponseDto feeLookupResponseDto) {
-        var feeAmountPounds = feeLookupResponseDto.getFeeAmount();
-
-        return feeAmountPounds.multiply(PENCE_PER_POUND).toBigInteger();
-    }
+    private Fee buildFeeDto(FeeLookupResponseDto feeLookupResponseDto) {
+        BigDecimal calculatedAmount = feeLookupResponseDto.getFeeAmount()
+            .multiply(PENCE_PER_POUND)
+            .setScale(0, RoundingMode.CEILING);
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/FeesService.java]: I think we should always use `RoundingMode.UNNECESSARY`. When there's going to be some weird value it's better to throw exception than charge more (even if it's only one pence).

File: src/main/java/uk/gov/hmcts/reform/unspec/service/OrganisationService.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    private final OrganisationApi organisationApi;
+    private final AuthTokenGenerator authTokenGenerator;
+
+    public Optional<Organisation> findOrganisation(String authToken) {
+        try {
+            return ofNullable(organisationApi.findUserOrganisation(authToken, authTokenGenerator.generate()));
+
+        } catch (FeignException.NotFound | FeignException.Forbidden ex) {
+            log.error("User not registered in MO", ex);
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/OrganisationService.java:28]: Is there some potentially useful data in response body? If yes, then it's better to log `ex.contentUTF8()`, because FeignException does not show it by default.

File: ccd-definition/CaseField.json (hunk 1 above)
Comment (swalker125) [reply ccd-definition/CaseField.json:159]: I'll be honest I wasn't sure about that from the get go, I think if naming and comments can explain a field sufficiently then that should do. However if @AlisonRevittHM asks for it then I will add.

File: e2e/pages/createClaim/paymentReference.page.js (hunk 2 above)
Comment (swalker125) [reply e2e/pages/createClaim/paymentReference.page.js:11]: To be honest, it isn't. The field is prepopulated from solicitor references, the clear was so that we type in new text

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java (hunk 4 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java]: much nicer

File: src/main/java/uk/gov/hmcts/reform/unspec/service/FeesService.java (hunk 5 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/FeesService.java]: sorry I think you mentioned this too me before, will update.

File: src/main/java/uk/gov/hmcts/reform/unspec/service/OrganisationService.java (hunk 6 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/OrganisationService.java:28]: Possible errors from this endpoint are:

400 | An invalid ID was provided
403 | Forbidden Error: Access denied
404 | No Organisation found with the given ID

We know that 400 in theory should never happen from a logged in user, 403 in theory should also never happen when we are whitelisted but the user could maybe not have the right roles to query this endpoint, so this just leaves the 404. I don't think we are swollowing anything useful there so will leave as is.

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java (hunk 3 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java:110]: I'm not sure how to get this working, the selected value is gone when I click back, maybe one to look at in tech debt?

Comment (pd-dudkiewicz) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java:110]: if it's not simple to fix then yeah, we could get that in separate ticket

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandlerTest.java [hunk 7]
```diff
@@ -141,6 +161,80 @@ void shouldReturnNoError_whenSoleTraderDateOfBirthIsInThePast() {
...
+                    feeData.getCode(),
+                    feeData.getVersion()
+                );
+
+            assertThat(getDynamicList(response))
+                .isEqualTo(DynamicList.builder().value(DynamicListElement.EMPTY).build());
+        }
+
+        private DynamicList getDynamicList(AboutToStartOrSubmitCallbackResponse response) {
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (pd-dudkiewicz) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandlerTest.java]: this could be moved to field, so we don't create new instance on every method call

**Review by pd-dudkiewicz — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java (hunk 3 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/CreateClaimCallbackHandler.java:110]: https://tools.hmcts.net/jira/browse/CMC-1122

File: ccd-definition/CaseField.json (hunk 1 above)
Comment (swalker125) [reply ccd-definition/CaseField.json:159]: This can always be retrofitted as well, has no functional effect for the story

**Review by pd-dudkiewicz — DISMISSED** [review]

**Review by pd-dudkiewicz — DISMISSED** [review]

**Review by pd-dudkiewicz — DISMISSED** [review]

**Review by pd-dudkiewicz — APPROVED** [review]

**Review by douglasrice — APPROVED** [review]
